File: ProjectTest/OneLang/CSharp/One/Transforms/FillAttributesFromTrivia.cs
using System.Collections.Generic;
using One.Ast;

namespace One.Transforms
{
    public class FillAttributesFromTrivia {
        public static Dictionary<string, string> processTrivia(string trivia) {
            var result = new Dictionary<string, string> {};
            if (trivia != null && trivia != "") {
                var regex = new RegExp("(?:\\n|^)\\s*(?:\\/\\/|#|\\/\\*\\*?)\\s*@([a-z0-9_.-]+) ?((?!\\n|\\*\\/|$).+)?");
                while (true) {
                    var match = regex.exec(trivia);
                    if (match == null)
                        break;
                    result.set(match.get(1), match.get(2) ?? "true");
                }
            }
            return result;
        }
        
        private static void process(IHasAttributesAndTrivia[] items) {
            foreach (var item in items)
                item.attributes = FillAttributesFromTrivia.processTrivia(item.leadingTrivia);
        }
        
        private static void processBlock(Block block) {
            if (block == null)
                return;
            FillAttributesFromTrivia.process(block.statements.ToArray());
            foreach (var stmt in block.statements) {
                if (stmt is ForeachStatement forStat)
                    FillAttributesFromTrivia.processBlock(forStat.body);
                else if (stmt is ForStatement forStat2)
                    FillAttributesFromTrivia.processBlock(forStat2.body);
                else if (stmt is IfStatement ifStat) {
                    FillAttributesFromTrivia.processBlock(ifStat.then);
                    FillAttributesFromTrivia.processBlock(ifStat.else_);
                }
            }
        }
        
        private static void processMethod(IMethodBaseWithTrivia method) {
            if (method == null)
                return;
            FillAttributesFromTrivia.process(new IMethodBaseWithTrivia[] { method });
            FillAttributesFromTrivia.processBlock(method.body);
        }
        
        public static void processFile(SourceFile file) {
            FillAttributesFromTrivia.process(file.imports);
            FillAttributesFromTrivia.process(file.enums);
            FillAttributesFromTrivia.process(file.interfaces);
            FillAttributesFromTrivia.process(file.classes);
            FillAttributesFromTrivia.processBlock(file.mainBlock);
            
            foreach (var intf in file.interfaces)
                foreach (var method in intf.methods)
                    FillAttributesFromTrivia.processMethod(method);
            
            foreach (var cls in file.classes) {
                FillAttributesFromTrivia.processMethod(cls.constructor_);
                FillAttributesFromTrivia.process(cls.fields);
                FillAttributesFromTrivia.process(cls.properties);
                foreach (var prop in cls.properties) {
                    FillAttributesFromTrivia.processBlock(prop.getter);
                    FillAttributesFromTrivia.processBlock(prop.setter);
                }
                foreach (var method in cls.methods)
                    FillAttributesFromTrivia.processMethod(method);
            }
        }
        
        public static void processPackage(Package pkg) {
            foreach (var file in Object.values(pkg.files))
                FillAttributesFromTrivia.processFile(file);
        }
    }
}

File: ProjectTest/OneLang/PHP/One/Transforms/FillAttributesFromTrivia.php
<?php

namespace One\Transforms\FillAttributesFromTrivia;

use One\Ast\Types\SourceFile;
use One\Ast\Types\IMethodBase;
use One\Ast\Types\IHasAttributesAndTrivia;
use One\Ast\Types\Package;
use One\Ast\Types\IMethodBaseWithTrivia;
use One\Ast\Statements\ForeachStatement;
use One\Ast\Statements\ForStatement;
use One\Ast\Statements\IfStatement;
use One\Ast\Statements\Block;
use One\Ast\Statements\WhileStatement;
use One\Ast\Statements\DoStatement;

class FillAttributesFromTrivia {
    static function processTrivia($trivia) {
        $result = Array();
        if ($trivia !== null && $trivia !== "") {
            $regex = new \OneLang\RegExp("(?:\\n|^)\\s*(?://|#|/\\*\\*?)\\s*@([a-z0-9_.-]+) ?((?!\\n|\\*/|$).+)?");
            while (true) {
                $match = $regex->exec($trivia);
                if ($match === null)
                    break;
                if (array_key_exists($match[1], $result))
                    $result[$match[1]] = @$result[$match[1]] ?? null . "\n" . $match[2];
                else
                    $result[$match[1]] = $match[2] ?? "true";
            }
        }
        return $result;
    }
    
    private static function process($items) {
        foreach ($items as $item)
            $item->attributes = FillAttributesFromTrivia::processTrivia($item->leadingTrivia);
    }
    
    private static function processBlock($block) {
        if ($block === null)
            return;
        FillAttributesFromTrivia::process($block->statements);
        foreach ($block->statements as $stmt) {
            if ($stmt instanceof ForeachStatement)
                FillAttributesFromTrivia::processBlock($stmt->body);
            else if ($stmt instanceof ForStatement)
                FillAttributesFromTrivia::processBlock($stmt->body);
            else if ($stmt instanceof WhileStatement)
                FillAttributesFromTrivia::processBlock($stmt->body);
            else if ($stmt instanceof DoStatement)
                FillAttributesFromTrivia::processBlock($stmt->body);
            else if ($stmt instanceof IfStatement) {
                FillAttributesFromTrivia::processBlock($stmt->then);
                FillAttributesFromTrivia::processBlock($stmt->else_);
            }
        }
    }
    
    private static function processMethod($method) {
        if ($method === null)
            return;
        FillAttributesFromTrivia::process(array($method));
        FillAttributesFromTrivia::process($method->parameters);
        FillAttributesFromTrivia::processBlock($method->body);
    }
    
    static function processFile($file) {
        FillAttributesFromTrivia::process($file->imports);
        FillAttributesFromTrivia::process($file->enums);
        FillAttributesFromTrivia::process($file->interfaces);
        FillAttributesFromTrivia::process($file->classes);
        FillAttributesFromTrivia::processBlock($file->mainBlock);
        
        foreach ($file->interfaces as $intf)
            foreach ($intf->methods as $method)
                FillAttributesFromTrivia::processMethod($method);
        
        foreach ($file->classes as $cls) {
            FillAttributesFromTrivia::processMethod($cls->constructor_);
            FillAttributesFromTrivia::process($cls->fields);
            FillAttributesFromTrivia::process($cls->properties);
            foreach ($cls->properties as $prop) {
                FillAttributesFromTrivia::processBlock($prop->getter);
                FillAttributesFromTrivia::processBlock($prop->setter);
            }
            foreach ($cls->methods as $method)
                FillAttributesFromTrivia::processMethod($method);
        }
    }
    
    static function processPackage($pkg) {
        foreach (array_values($pkg->files) as $file)
            FillAttributesFromTrivia::processFile($file);
    }
}
